### ISSUE TITLE
Question about SSR and Service Workers for SEO

### ISSUE THREAD

**Issue by zhandosweb** [gatsbyjs/gatsby#20235]
Hello!

Please, let me ask one question.

I read the documentation. And there was written, that if I `Right-Click` and `View Source`, I can't see HTML-tags, like headlines, paragraphs, etc, because there are servise-workers do some jobs, like DOM-cache.

But how Google can see actual HTML-DOM? Gatsby can identify Google-bot using User Agent? Or how it works? 

Sorry, sir, for bad English. I'm still learning. Thank you for your work! I'm a newbie, but I really fell in love with Gatsby!

**Comment by wardpeet** [comment]
Hey, thanks for reaching out!

We should definitely explain this somewhere in our docs so the fault is all ours and not yours. Thank you for raising these questions as more people will have it.

A service-worker is installed during the page-load so the first page-view that a user gets will be the full HTML page. On the next loads, they will receive these "empty" pages like you're seeing. You may ask why? There is no need for us to store the whole HTML structure as we could just save the assets and pageData. This allows us to reduce the bandwidth used for these loads and also lower memory costs on user's devices.

Back to your original question, Google bot will always visit your page with a fresh slate so they will get the full HTML instead of the empty page. You do get ranked better when using a service worker as Google will see that you're optimizing your page for users.

If you want to view it for yourself the best way is to use 
linux and mac: `curl <url>`
windows: `Invoke-WebRequest <url> | Select -ExpandProperty Content`

See https://www.gatsbyjs.org/docs/seo/#server-rendering

**Comment by zhandosweb** [comment]
Thank you very much for your answer! 

Yes, I already read the documentation. And checked the source code using `curl`, and there is see all HTML-tags. 

If Google-bot visits my website, he gets clean HTML DOM. But how about other search engines? I do SEO for my sites in Russia. And the main search engine for us - Yandex, and the second is Google. Other search engines (bots) also get clean DOM HTML? 

And sir, let me ask the last question. I found a good Gatsby SEO Guide, here is a link: https://gregberge.com/blog/gatsby-seo. And if I check source code in the browser (`ctrl + u`), I see all HTML-tags, and service-workers not hiding them. What do you think about it?

**Comment by wardpeet** [comment]
This is mainly how SEO bots work (also goes for facebook, twitter when sharing a site): they crawl your website HTML basically like curl or invoke-webrequest works (there is more too it but basically it this). Google (and probably others) also do a real site check using a browser. All these checks are run in sandbox/isolated mode so it's always fresh and the full html.

You can test it ourself by opening your site in incognito mode and do a view-source (make sure no other incognito tab is open).

The gregberge.com isn't using a service worker so it's normal that it just returns the same HTML. There is nothing wrong with it but a service worker will make your site faster as on a next visit, all resources will be there even if you lose connection. Think about when you're traveling and you go through a tunnel or just have bad connection, it will still work.

So in short, everything is fine ;)

**Comment by zhandosweb** [comment]
Thank you very much for the answers! Now I understand everything. 

And yes, if I view the source code of the site in an incognito tab, I can see all HTML DOM. 

Gatsby is amazing, I love it!!! #$uck PHP and Wordpress. I close the issue, no have any questions more. 

Thank you again for the answers!
